### PR TITLE
feat(logger): introduce log key reordering functionality

### PR DESCRIPTION
**Pull request by arnabrahman**
## Summary
By default, we can not change the log key positions. This PR gives the user the ability to change log key positions. We can change log order positions for any keys added in runtime
### Changes

- Mostly followed the recommendations from the [comment](https://github.com/aws-powertools/powertools-lambda-typescript/issues/1568#issuecomment-1786186089) from the actual issue.
- A new option ` logRecordOrder` is added for reordering.
- Change `formatAttributes` function for the new option.
- Write unit tests for this.
- Necessary documentation has been added with examples

**Issue number:** #1568 

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by dreamorosi** [comment]
Hi @arnabrahman, thank you for opening this PR!

I'll be reviewing it this morning and will get back to you, but overall I think we're going in the right direction.

File: packages/logger/src/formatter/PowertoolsLogFormatter.ts
Comment (dreamorosi) [inline packages/logger/src/formatter/PowertoolsLogFormatter.ts:1]: I've been playing around this, and since we are not doing any kind of deep object merge, I think we can try to implement it without using Lodash.

I've created this test bench suite ([here](https://jsbench.me/kslycmt0hc/1)) to compare the current implementation with some alternatives, and it looks like this one is ~35% faster:

```ts
const attributes = {};

for (const key of logRecordOrder) {
  if (key in baseAttributes && !(key in attributes)) {
    attributes[key] = baseAttributes[key];
  }
}

for (const key in baseAttributes) {
  if (!(key in attributes)) {
		attributes[key] = baseAttributes[key];
	}
}
```

**Comment by dreamorosi** [comment]
> i believe we should type the option for standard structured keys

I was under the same impression, but looking at the docs again I think we're mistaken, since the docs mention this:

> You can change the order of [standard Logger keys](https://docs.powertools.aws.dev/lambda/python/latest/core/logger/#standard-structured-keys) or any keys that will be appended later at runtime

Additionally, in the example they show usage of a key (`request_id`) that is not part of the standard keys and that is added at runtime.

With this in mind I think we should follow the same pattern, if possible.

I haven't tested this in the code, but at first glance if I remember correctly, I think we should be able to include all keys to the new logic by working with the `additionalLogAttributes` (of type `LogAttributes`) object.

Doing this will mean we probably remove this line, which I'm not sure if it'll have any impact on the log output.

https://github.com/aws-powertools/powertools-lambda-typescript/blob/d76eb1d3d1569bd4675f2ff74c909b5b3976e232/packages/logger/src/formatter/PowertoolsLogFormatter.ts#L39

If this is a viable option, we should also modify the types of the new `logRecordOrder` accordingly (maybe an union between the keys of `UnformattedAttributes` and `LogAttributes`?).

---

Let me know if I cleared your questions 😄

**Comment by arnabrahman** [comment]
This summer felt like an eternity but good to be back. 

I was looking at your suggestions @dreamorosi, on a high level it should be possible to include all the `additional` keys although I have to look a bit more to validate this. But first I have to merge the `main` branch but I cannot. I am getting `pre-commit` error for `markdownlint-cli2` in `packages/tracer/README.md`. I ran `setup-local` & tried again but still got the same result. Is there anything that needs to be done?

Screenshot from 2024-08-25 16-19-59

**Comment by arnabrahman** [comment]
@dreamorosi So i have progressed on this. But I have another question. Right now the `logRecordOrder` passes down to child logger. What if we want to override `logRecordOrder` to child instance?


Currently while creating child, `logFormatter` is passed to child instance. https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/packages/logger/src/Logger.ts#L308

And `logFormatter` is not reiniliazed if it's already set https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/packages/logger/src/Logger.ts#L1085


In this case, if we provide `logRecordOrder` to child, it will not be overridden. Correct me if I am wrong

**Comment by dreamorosi** [comment]
> In this case, if we provide logRecordOrder to child, it will not be overridden. Correct me if I am wrong

You are correct, we reuse the same log formatter from the parent because up until now it was reasonable to do so.

For the time being I would opt for keeping the logic as-is and prevent customers from overriding `logRecordOrder` when creating a child. 

This means that child loggers will have the same log formatter and log record order as their parent. If there's demand for this feature we can easily add it in the future, but for now I'd simply go ahead as-is.

**Comment by dreamorosi** [comment]
Hi @arnabrahman, thank you for the PR - I've started reviewing it and should be able to finish by tomorrow morning at the latest.

File: packages/logger/src/formatter/LogFormatter.ts [hunk 1]
```diff
@@ -19,8 +19,14 @@ abstract class LogFormatter implements LogFormatterInterface {
...
   protected envVarsService?: EnvironmentVariablesService;
 
+  /**
+   * An array of keys that defines the order of the log record.
+   */
+  protected logRecordOrder?: LogRecordOrder;
+
   public constructor(options?: LogFormatterOptions) {
     this.envVarsService = options?.envVarsService;
+    this.logRecordOrder = options?.logRecordOrder;
```
Comment (dreamorosi) [inline packages/logger/src/formatter/LogFormatter.ts:29]: I've been thinking about this a bit (also why the review took a bit longer), and I think this property - and feature - should be only available in the `PowertoolsLogFormatter` class.

My reasoning is that if customers are creating their own log formatter class by extending `LogFormatter`, then they already have full control over the order of the keys - so having this available doesn't add much and instead pollutes the API.

I think we should position this feature as a way to lightly extend the default log formatter without having to create your own. This also matches what the Python implementation of Powertools does.

To sum up, I would move the `logRecordOrder` property in the `PowertoolsLogFormatter` and use it only there. Additionally, I would also extend the `ConstructorOptions` type to make `logRecordOrder` & `logFormatter` mutually exclusive, [similar to what we do for this other use case](https://github.com/aws-powertools/powertools-lambda-typescript/blob/23a33dc52db9d2bb8cfa8ff043abcc1020ff958e/packages/logger/src/types/Logger.ts#L131).

**Review by dreamorosi — CHANGES_REQUESTED** [review]
Hi @arnabrahman, thank you so much for the patience while I reviewed this and also for the great work!

I think we are getting very closer, I've left a few comments but the main implementation is done I think. 

Let me know if you have any questions or disagree with the comments.

**Review by dreamorosi — APPROVED** [review]
Thank you @arnabrahman for another great addition and contribution.

I really appreciate you coming back to this and working on the various iterations until we got here.

Thank you again!

PS: I'll reply to your comment on the other issue later today 😉
